File: d03/d03.py
def most_common(l, j):
    return int(2 * sum(l[i][j] for i in range(len(l))) >= len(l))

def iter(l, j, least = False):
    which = most_common(l, j) ^ least
    return [s for s in l if s[j] == which]

def to_int(l):
    return int(''.join(map(str,l)), 2)

def star1(inp):
    s = to_int([most_common(inp, j) for j in range(cols)])
    neg_s = (1 << cols) - 1 - s
    return s * neg_s

def star2(inp):
    nums_most, nums_least = inp.copy(), inp.copy()
    for j in range(cols):
        nums_most = iter(nums_most, j)
        nums_least = iter(nums_least, j, least = len(nums_least) > 1)

    return to_int(nums_most[0]) * to_int(nums_least[0])


inp = [[*map(int, line[:-1])] for line in open('input.txt')]
cols = len(inp[0])

print('Star 1:', star1(inp))
print('Star 2:', star2(inp))

File: d03/d03.hs
import qualified Data.Sequence as S
import Data.Bits

mostCommon :: [S.Seq Int] -> Int -> Int
mostCommon xs j = fromEnum (2 * s >= n)
  where
      n = length xs
      s = sum [x `S.index` j | x <- xs]

iter :: Int -> [S.Seq Int] -> Int -> [S.Seq Int]
iter least xs j = filter (\x -> x `S.index` j == which) xs
  where
      which = mostCommon xs j `xor` least

toInt :: S.Seq Int -> Int
toInt = foldl (\acc x -> 2 * acc + x) 0

star1 :: [S.Seq Int] -> Int
star1 xs = x * negX
  where
      m = S.length $ head xs
      x = toInt . S.fromList . foldr (\x acc -> (mostCommon xs x):acc) [] $ [0..m-1]
      negX = (shift 1 m) - 1 - x

star2 :: [S.Seq Int] -> Int
star2 xs = (res 0) * (res 1)
  where
      m = S.length $ head xs
      f least acc x
          | length acc > 1 = iter least acc x
          | otherwise      = acc
      res least = toInt . head . foldl (f least) xs $ [0..m-1]

main = do
    inp <- readFile "input.txt"
    let l = [ S.fromList [read [c] | c <- s] | s <- lines inp]
    putStrLn $ "Star 1: " ++ (show . star1 $ l)
    putStrLn $ "Star 2: " ++ (show . star2 $ l)
